File: apps/ios/Sources/Core/Media/MediaUpload.swift
import Foundation

struct PendingMedia: Codable, Sendable, Equatable { let assetId: String; let kind: MediaKind }
// Journal owner must atomically check original scope at database commit. No signed URLs/tokens.
protocol MediaJournal: Sendable {
    func save(_ pending: PendingMedia, scope: any MediaScope) async throws
    func remove(_ assetID: String, scope: any MediaScope) async throws
}
enum MediaUploadState: Sendable {
    case idle, reserving, uploading(PendingMedia), processing(PendingMedia)
    case ready(PendingMedia, MediaReceipt), failed(PendingMedia?), cancelled
}
@MainActor final class MediaUpload {
    private let client: MediaClient
    private let journal: any MediaJournal
    private(set) var state: MediaUploadState = .idle
    init(client: MediaClient, journal: any MediaJournal) { self.client = client; self.journal = journal }
    func start(_ file: MediaFile) async throws -> MediaReceipt {
        switch state { case .reserving, .uploading, .processing: throw MediaError.invalid; default: break }
        var pending: PendingMedia?
        defer { file.remove() }
        do {
            try client.scope.check(); state = .reserving
            let receipt = try await client.reserve(file)
            let record = PendingMedia(assetId: receipt.assetId, kind: file.kind); pending = record
            try await journal.save(record, scope: client.scope); try client.scope.check()
            state = .uploading(record)
            _ = try await client.upload(record.assetId, file: file)
            state = .processing(record)
            let ready = try await client.awaitReady(record.assetId)
            try client.scope.check(); state = .ready(record, ready); return ready
        } catch { state = error is CancellationError || Task.isCancelled ? .cancelled : .failed(pending); throw error }
    }
    // Restart recovery only polls. Unknown admission must never cause an automatic upload replay.
    func recover(_ pending: PendingMedia) async throws -> MediaReceipt {
        switch state { case .reserving, .uploading, .processing: throw MediaError.invalid; default: break }
        do {
            try client.scope.check(); state = .processing(pending)
            let ready = try await client.awaitReady(pending.assetId)
            try client.scope.check(); state = .ready(pending, ready); return ready
        } catch { state = error is CancellationError || Task.isCancelled ? .cancelled : .failed(pending); throw error }
    }
    // Call after the owning send/profile coordinator durably accepted the server acknowledgement.
    func acknowledged(_ assetID: String) async throws {
        try client.scope.check(); try await journal.remove(mediaID(assetID), scope: client.scope)
    }
}

File: apps/android/app/src/main/java/chat/rogi/rogichat/core/media/MediaUpload.kt
package chat.rogi.rogichat.core.media

import kotlinx.coroutines.CancellationException

/** Existing scoped persistence supplies this journal; never persist signed URLs or credentials.
 * save/remove must commit under the same original scope, not the currently logged-in account. */
interface MediaJournal {
    suspend fun save(scope: MediaScope, pending: PendingMedia)
    suspend fun remove(scope: MediaScope, assetId: String)
}
data class PendingMedia(val assetId: String, val kind: MediaKind) { init { mediaId(assetId) } }
sealed interface MediaUploadState {
    data object Idle : MediaUploadState
    data object Reserving : MediaUploadState
    data class Uploading(val pending: PendingMedia) : MediaUploadState
    data class Processing(val pending: PendingMedia) : MediaUploadState
    data class Ready(val pending: PendingMedia, val receipt: MediaReceipt) : MediaUploadState
    data class Failed(val pending: PendingMedia?, val retryStatusOnly: Boolean) : MediaUploadState
    data object Cancelled : MediaUploadState
}
/** Serial operation owned by a lifecycle job. A lost upload response is recovered through status,
 * never a blind POST replay. Local scratch is deleted on success, error and cancellation. */
class MediaUpload(private val client: MediaClient, private val journal: MediaJournal) {
    var state: MediaUploadState = MediaUploadState.Idle; private set
    suspend fun start(file: MediaFile): MediaReceipt {
        check(state !is MediaUploadState.Reserving && state !is MediaUploadState.Uploading && state !is MediaUploadState.Processing)
        var pending: PendingMedia? = null
        try {
            client.scope.check(); state = MediaUploadState.Reserving
            val reservation = client.reserve(file)
            pending = PendingMedia(reservation.assetId, file.kind)
            journal.save(client.scope, pending); client.scope.check()
            state = MediaUploadState.Uploading(pending)
            client.upload(pending.assetId, file)
            state = MediaUploadState.Processing(pending)
            val receipt = client.awaitReady(pending.assetId)
            client.scope.check(); state = MediaUploadState.Ready(pending, receipt)
            return receipt
        } catch (error: CancellationException) { state = MediaUploadState.Cancelled; throw error }
        catch (error: Exception) { state = MediaUploadState.Failed(pending, pending != null); throw error }
        finally { file.close() }
    }
    suspend fun recover(pending: PendingMedia): MediaReceipt {
        check(state !is MediaUploadState.Reserving && state !is MediaUploadState.Uploading && state !is MediaUploadState.Processing)
        try {
            client.scope.check(); state = MediaUploadState.Processing(pending)
            val receipt = client.awaitReady(pending.assetId)
            client.scope.check(); state = MediaUploadState.Ready(pending, receipt); return receipt
        } catch (error: CancellationException) { state = MediaUploadState.Cancelled; throw error }
        catch (error: Exception) { state = MediaUploadState.Failed(pending, true); throw error }
    }
    /** Call only after server SEND/profile acknowledgement was persisted by its owning coordinator. */
    suspend fun acknowledged(assetId: String) { client.scope.check(); journal.remove(client.scope, mediaId(assetId)) }
}
